### PR TITLE
don't add pod's configmap volume and secret volume to desiredStateOfWord when the pod's status is Terminating

### PR DESCRIPTION
**Pull request by chenyw1990**
What type of PR is this?
/kind bug

What this PR does / why we need it:
If we delete the pod and the configmap used by the pod when kubelet is stoped, after kubelet started, the configmap volume will mount failed forever. so we should not add the configmap volume to desiredStateOfWorld when pod's DeletionTimestamp is not empty.

If we delete the pod and the secret used by the pod when kubelet is stoped, after kubelet started, the secret volume will mount failed forever. so we should not add the secret volume to desiredStateOfWorld when pod's DeletionTimestamp is not empty.

Which issue(s) this PR fixes:
#96635

Special notes for your reviewer:

Does this PR introduce a user-facing change?:
NONE

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

### CODE REVIEW

**Comment by chenyw1990** [comment]
@liggitt @mborsz

File: pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go [hunk 1]
```diff
@@ -332,6 +332,20 @@ func (dswp *desiredStateOfWorldPopulator) processPodVolumes(
 			continue
 		}
 
+		// If we delete the pod and the configmap used by the pod when kubelet is stoped, after kubelet started, the configmap volume will mount failed forever.
+		// so we should not add the configmap volume to desiredStateOfWorld when pod's DeletionTimestamp is not empty.
+		if pod.DeletionTimestamp != nil && podVolume.VolumeSource.ConfigMap != nil {
```
Comment (liggitt) [inline pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:337]: Special-casing secret/configmap volumes here seems strange. What about:
* projected volumes sourced from secret/configmap objects
* projected volumes sourced from service accounts
* PVC volumes whose PVC object has been removed

Does this also assume termination will be short and it is ok to stop updating secret/configmap volumes during that time? Is that a good assumption (containers can take significant time to terminate)?

Is existing secret/configmap content left as-is if we omit these or is it removed? If it is removed, can that break a container while it is terminating?

**Comment by msau42** [comment]
/assign @jingxu97 @gnufied 
@kubernetes/sig-storage-pr-reviews 

I remember we fixed some secret/configmap mounting issues in the past when the secret/configmap object didn't exist

**Comment by gnufied** [comment]
Looks similar to https://github.com/kubernetes/kubernetes/pull/94057/files

**Comment by mborsz** [comment]
What is a state of this PR?

**Comment by SergeyKanzhelev** [comment]
Seems like there is an outstanding comment from @liggitt  triaged to the waiting for author column

**Comment by chenyw1990** [comment]
I tested the latest code in master, this problem no longer exists, I am looking for the reason

**Comment by marseel** [comment]
@chenyw1990 Do you have any update?
I've added new test run to issue which shows that this problem still occurs.

**Comment by mm4tt** [comment]
Yeah, the problem still occurs. A 5k node performance tests has run into this a few days ago - https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1381291447344435200

Any updates?

**Comment by chenyw1990** [comment]
@mm4tt 
I have test again with kubelet version Kubernetes v1.22.0-alpha.0.3+770d3f181c5d7e-dirty
test yaml is ：

```

kind: ConfigMap
apiVersion: v1
metadata:
  name: configmap-test
data:
  test: |
    test
---  
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test
  labels:
    app: test
spec:
  selector:
    matchLabels:
      app: test
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 50%
      maxUnavailable: 0
    type: RollingUpdate
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - name: test1
        image: docker.io/library/nginx
        imagePullPolicy: IfNotPresent
        command: ["sleep", "1000000"]
        resources:
          requests:
            cpu: 5m
            memory: 10Mi
        ports:
        - containerPort: 81
          name: test
        volumeMounts:
        - mountPath: /configmap-test
          name: configmap-test
          subPath: test
      volumes:
      - name: configmap-test
        configMap:
          defaultMode: 256
          name: configmap-test
          optional: false
          items: 
          - key: test
            path: test
            mode: 256
```
when kubelet is stoped, delete configmap, and scale down delployment test, after kubelet been started, pod will been deleted.
image

but the pod volume dir in /var/lib/kubelet/pods/1df78fe4-5302-493e-8966-cec417720dcd/volumes not been cleared
image

I don't know why.

pod is test-f68c6d4d6-j9bqc_default(1df78fe4-5302-493e-8966-cec417720dcd)
the kubelet' log is :

[kubelet.log](https://github.com/kubernetes/kubernetes/files/6304503/kubelet.log)

**Comment by jingxu97** [comment]
> @mm4tt
> I have test again with kubelet version Kubernetes v1.22.0-alpha.0.3+770d3f181c5d7e-dirty
> test yaml is ：
> 
> ```
> 
> kind: ConfigMap
> apiVersion: v1
> metadata:
>   name: configmap-test
> data:
>   test: |
>     test
> ---  
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: test
>   labels:
>     app: test
> spec:
>   selector:
>     matchLabels:
>       app: test
>   replicas: 1
>   strategy:
>     rollingUpdate:
>       maxSurge: 50%
>       maxUnavailable: 0
>     type: RollingUpdate
>   template:
>     metadata:
>       labels:
>         app: test
>     spec:
>       containers:
>       - name: test1
>         image: docker.io/library/nginx
>         imagePullPolicy: IfNotPresent
>         command: ["sleep", "1000000"]
>         resources:
>           requests:
>             cpu: 5m
>             memory: 10Mi
>         ports:
>         - containerPort: 81
>           name: test
>         volumeMounts:
>         - mountPath: /configmap-test
>           name: configmap-test
>           subPath: test
>       volumes:
>       - name: configmap-test
>         configMap:
>           defaultMode: 256
>           name: configmap-test
>           optional: false
>           items: 
>           - key: test
>             path: test
>             mode: 256
> ```
> 
> when kubelet is stoped, delete configmap, and scale down delployment test, after kubelet been started, pod will been deleted.
> image
> 
> but the pod volume dir in /var/lib/kubelet/pods/1df78fe4-5302-493e-8966-cec417720dcd/volumes not been cleared
> image
> 
> I don't know why.
> 
> pod is test-f68c6d4d6-j9bqc_default(1df78fe4-5302-493e-8966-cec417720dcd)
> the kubelet' log is :
> 
> [kubelet.log](https://github.com/kubernetes/kubernetes/files/6304503/kubelet.log)

I found the following error from kubelet log
```
Operation for "{volumeName:kubernetes.io/configmap/1df78fe4-5302-493e-8966-cec417720dcd-configmap-test podName:1df78fe4-5302-493e-8966-cec417720dcd nodeName:}" failed. No retries permitted until 2021-04-13 14:21:54.356972357 +0000 UTC m=+17.937967176 (durationBeforeRetry 4s). Error: "MountVolume.SetUp failed for volume \"configmap-test\" (UniqueName: \"kubernetes.io/configmap/1df78fe4-5302-493e-8966-cec417720dcd-configmap-test\") pod \"test-f68c6d4d6-j9bqc\" (UID: \"1df78fe4-5302-493e-8966-cec417720dcd\") : configmap \"configmap-test\" not found"
```

But since configmap is not found, not sure why volume dir is created..

**Comment by ehashman** [comment]
Tests aren't passing, @wojtek-t if you want to revive this PR you may want to open a new one. I think the issue is that we haven't actually been able to work out a reproducer, and hence we have no idea if the fix is working.

**Comment by dims** [comment]
@chenyw1990 ping!

**Pull request closed without merging**